stan/math/rev/mat/fun/multiply.hpp

```diff
@@ -1,23 +1,407 @@
+                Bd_a[i] = B.coeffRef(i).val();
+            }
+            Eigen::Matrix<double, RA_, CB_> AB
+              = Eigen::Map<Eigen::Matrix<double, RA_, CA_> >(Ad_a, RA, CA)
+              * Eigen::Map<Eigen::Matrix<double, CA_, CB_> >(Bd_a, CA, CB);
+            for (size_type i = 0; i < AB.size(); ++i)
+                variRefAB_[i] = new vari(AB.coeffRef(i), false);
+        }
+
+      virtual void chain() {
+        using Eigen::Matrix;
+        using Eigen::Map;
+        Matrix<double, RA_, CB_> adjAB(RA, CB);
+        Matrix<double, RA_, CA_> adjA(RA, CA);
+        Matrix<double, CA_, CB_> adjB(CA, CB);
+
+        for (size_type i = 0; i < adjAB.size(); ++i)
+            adjAB(i) = variRefAB_[i]->adj_;
+        adjA = adjAB * Map<Matrix<double, CA_, CB_> >(Bd_a, CA, CB).transpose();
+        adjB = Map<Matrix<double, RA_, CA_> >(Ad_a, RA, CA).transpose() * adjAB;
+        for (size_type i = 0; i < Asize; ++i)
+            variRefA_[i]->adj_ += adjA(i);
+        for (size_type i = 0; i < Bsize; ++i)
+            variRefB_[i]->adj_ += adjB(i);
+      }
+    };
+
+    template <typename TA, int CA_, typename TB>
+    class multiply_mat_vari<TA, 1, CA_, TB, 1> : public vari {
+    public:
+      int CA;
+      double* Ad_a;
+      double* Bd_a;
+      vari** variRefA_;
+      vari** variRefB_;
+      vari* variRefAB_;
+
+      /* ctor for multiply function
+       *
+       * Stores pointers to varis for A
+       * Stores pointers to varis for B
+       * Stores pointers to varis for AB
+       * All memory allocated on ChainableStack
+       *
+       * @tparam TA Scalar type for matrix A
+       * @tparam CA_ Columns for matrix A, Rows for matrix B
+       * @tparam TB Scalar type for matrix B
+       * @param matrix A
+       * @param matrix B
+       * */
+     multiply_mat_vari(const Eigen::Matrix<TA, 1, CA_>& A,
+                       const Eigen::Matrix<TB, CA_, 1>& B)
+        : vari(0.0),
+          CA(A.cols()),
+          Ad_a(ChainableStack::memalloc_.alloc_array<double>
+               (A.cols())),
+          Bd_a(ChainableStack::memalloc_.alloc_array<double>
+               (B.rows())),
+          variRefA_(ChainableStack::memalloc_.alloc_array<vari*>
+                    (A.rows() * A.cols())),
+          variRefB_(ChainableStack::memalloc_.alloc_array<vari*>
+                    (B.rows() * B.cols())) {
+            for (size_type i = 0; i < A.size(); ++i) {
+                variRefA_[i] = A.coeffRef(i).vi_;
+                Ad_a[i] = A.coeffRef(i).val();
+            }
+            for (size_type i = 0; i < B.size(); ++i) {
+                variRefB_[i] = B.coeffRef(i).vi_;
+                Bd_a[i] = B.coeffRef(i).val();
+            }
+            double AB = Eigen::Map<
+              Eigen::Matrix<double, 1, CA_> > (Ad_a, 1, CA)
+                        * Eigen::Map<
+              Eigen::Matrix<double, CA_, 1> > (Bd_a, CA, 1);
+            variRefAB_ = new vari(AB, false);
```

Comment:
This is too hard to read.  Please either typedef these things (and in this case, I think it's just `RowVectorXd` and `VectorXd`, but I could be wrong), or format them to make them easier to read.  I'd do it as

```
using Eigen::Map;
using Eigen::Matrix;
double AB = Map<Matrix<double, 1, CA_> >(Ad_a, 1, CA)
  * Matrix<double, CA_, 1> >(Bd_a, CA, 1);
```

Note the lack of space before the `(`.  I think that space is an R thing because I see it from Andrew and many of our users' Stan code.